lib/Dialect/HW/Transforms/HWSpecialize.cpp

```diff
@@ -307,9 +306,7 @@ void HWSpecializePass::runOnOperation() {
       }
 
       // Extend the symbol cache with the newly created module.
-      sc.unfreeze();
       sc.addDefinition(specializedModule.getNameAttr(), specializedModule);
-      sc.freeze();
```

Comment:
So this removes the explicit freeze/unfreeze API for the base SymbolCache?

Reply:
correct

Reply:
In that case, one nitpick would be to keep the Mutable name for the base, since that seemed like an important distinction. More broadly, why did we have that API before? If it was valuable, why is it being removed now? Sorry for the late questioning here, but this is just a generic line of reasoning that I have for this sort of change.

Reply:
In HW, the symbol cache is being used in a multithreaded context, hence the need for freezing to assert for safety. In terms of it being valuable, I'd rather see the base case to be a simple "dumb" lookup table with no added semantics, and if more elaborate safety is needed, users can specialize to their liking (As in `HWSymbolCache`). I'm not strongly against keeping the Mutable name, but i don't think it is strictly necessary; we don't have `MutableDenseMap,...` either and instead assume that the programmer are aware of the restrictions of the datastructure.

Reply:
I guess my point not just the naming, but that the mutability/immutability is being flipped. Before the symbol cache was threadsafe by default, avoiding one class of potential footguns, and was extended to make it mutable when needed. Now it seems like the default is mutable, and that is extended to be made immutable. I guess my concern here is the base is now the less safe option than before. I don't know if this is worth changing, but that's what I'm trying to get at here.

Reply:
hmm... I understand. I think I'd still be in favor of keeping things like in the PR - to me it seems natural to have base types in a class hierarchy less restricted (=> less safe (?)) and let restrictions (safety(?)) increase with inheritance. And as such, while it is a flipping of the default behavior, it's flipping to somewhat more "expected" behavior (anecdotally, when I first worked with the `SymCache`, the freeze/unfreeze behavior was unexpected to me and seemed a strange restriction for what I thought the cache ought to have done and been).

Reply:
I'm not against changing it in this PR, and your argument makes sense to me.

Reply:
I don't think renaming the base class to mutable makes sense since implementations extending it can be either mutable, freeze-able, or immutable.

Yeah, we tried freeze-able in the base and it was pretty awkward. This is more natural to me -- if you need it to be freeze-able, just instantiate the freeze-able variant and pass that around. IMO the base class is providing a standard API to add symbols which I could see being useful in many cases.